FunctionalEquivalence/scripts/f1_evaluation.py

```diff
@@ -0,0 +1,253 @@
+import numpy as np
+import matplotlib
+import matplotlib.pyplot as plt
+import csv
+import statistics as stat
+import gzip
+import argparse
+
+matplotlib.rcParams['text.usetex'] = False
+matplotlib.rcParams['mathtext.default'] = 'regular'
+matplotlib.rcParams['font.family'] = 'serif'
+
+def f1(tp, fp, fn):
+    return tp / (tp + 0.5 * (fp + fn))
+
+def read_roc_to_f1_dict(filename):
+    f1_mean_dict = dict()
+    f1_list_dict = dict()
+    with gzip.open(filename, 'rt') as roc_file:
+        for line in roc_file:
+            if line.startswith('#score field'):
+                break
+        for line in csv.DictReader(roc_file, delimiter='\t'):
+            f1_score = f1(float(line['true_positives_baseline']), float(line['false_positives']), float(line['false_negatives']))
+            qual = round(float(line['#score']))
+            if qual not in f1_list_dict:
+                f1_list_dict[qual] = [f1_score]
+            else:
+                f1_list_dict[qual].append(f1_score)
+    
+    for qual, f1_list in f1_list_dict.items():
+        f1_mean_dict[qual] = stat.mean(f1_list)
+                                    
+    return f1_mean_dict
+
+def read_file(data, filename):
+    basename = filename[filename.rfind('/') + 1:]
+    info = basename.split('_')
+    evalinfo = info[0].split('.')
+    dataset = evalinfo[0]
+    replicate = evalinfo[1]
+    tool = evalinfo[2]
+    
+    if info[2] == 'vcfeval':
+        region = 'all'
+    else:
+        region = info[2]
+    if info[3 if region == 'all' else 4] == 'non':
+        var_type = 'indel'
+    else:
+        var_type = 'snp'
+    
+    f1_data = read_roc_to_f1_dict(filename)
+    
+    if dataset not in data:
+        data[dataset] = dict()
+    if replicate not in data[dataset]:
+        data[dataset][replicate] = {'tool1': {'all': dict(), 'HCR': dict(), 'LCR': dict()},
+                                    'tool2': {'all': dict(), 'HCR': dict(), 'LCR': dict()}, }
+    data[dataset][replicate][tool][region][var_type] = f1_data
+
+def extract_data_from_dict(score_range, data, dataset, tool, region, var_type):
+    return [
+                np.array([
+                    data[dataset][replicate][tool][region][var_type][score] if score in data[dataset][replicate][tool][region][var_type] else np.nan
+                    for replicate in data[dataset].keys()
+                ])
+                for score in score_range
+            ]
+    
+
+
+def aggregate_data(data, dataset, signed_difference):
```

Comment:
I know what you mean, can we maybe table this improvement for now? I thought about it for a while and I'm pretty sure it would require a rewrite of most of the code, because I'm relying on integers for indexing. I agree that it's worth it, but rewriting and especially testing it for correctness will take a while.

Reply:
Created issue for this: https://github.com/broadinstitute/palantir-workflows/issues/48